### PR TITLE
Add a timeout option to gtest-parallel.

### PR DESCRIPTION
**Pull request by ehlemur-zz**
Stop the tests' execution once the timeout has been exceeded.

### CODE REVIEW

File: gtest-parallel [hunk 1]
```diff
@@ -58,6 +58,9 @@ class SigintHandler(object):
   def __sigint_handler(self, signal_num, frame):
     with self.__lock:
       self.__on_sigint()
+  def interrupt(self):
```
Comment (ehlemur-zz) [inline gtest-parallel]: Maybe we could get rid of __sigint_handler and use this method instead?
Is interrupt a good name?

**Comment by ehlemur-zz** [comment]
@pbos @kwiberg

File: gtest-parallel [hunk 2]
```diff
@@ -465,7 +492,12 @@ def start_daemon(func):
 
 workers = [start_daemon(worker) for i in range(options.workers)]
 
-[t.join() for t in workers]
+try:
+  timeout.start()
```
Comment (kwiberg) [inline gtest-parallel]: For paranoia reasons, it's probably better to start the timeout before you start the worker threads. So IOW, maybe move the "workers =" line into the try...except, below this line?

File: gtest-parallel [hunk 3]
```diff
@@ -314,6 +334,9 @@ parser.add_option('--dump_json_test_results', type='string', default=None,
                   help='Saves the results of the tests as a JSON machine-'
                        'readable file. The format of the file is specified at '
                        'https://www.chromium.org/developers/the-json-test-results-format')
+parser.add_option('--timeout', type='int', default=None,
+                  help='Interrupt all remaining processes after a given '
+                       'timeout.')
```
Comment (kwiberg) [inline gtest-parallel]: Document what the integer means. Seconds?

File: gtest-parallel [hunk 4]
```diff
@@ -272,6 +291,7 @@ class TestTimes(object):
     except IOError:
       pass  # ignore errors---saving the times isn't that important
 
+
```
Comment (kwiberg) [inline gtest-parallel]: Spurious newline?

File: gtest-parallel (hunk 1 above)
Comment (kwiberg) [reply gtest-parallel]: Yes, interrupt is a good name. But __sigint_handler is registered as a sigint handler, so if you eliminated it you'd have to... yes, you could register a lambda as sigint handler instead; the lambda would just ignore the signal_num and frame arguments and call interrupt(). Do that.

File: gtest-parallel [hunk 5]
```diff
@@ -168,11 +183,9 @@ class FilterFormat:
     stdout_lock.release()
 
   def end(self):
-    if self.failures:
-      self.out.permanent_line("FAILED TESTS (%d/%d):"
-                              % (len(self.failures), self.total_tests))
-      for (binary, test) in self.failures:
-        self.out.permanent_line(" " + binary + ": " + test)
+    self.print_tests("FAILED TESTS", self.failures)
+    interruptions = set(self.started).difference(set(self.finished))
```
Comment (kwiberg) [inline gtest-parallel]: Can also be written as
`set(self.started) - set(self.finished)`

File: gtest-parallel [hunk 6]
```diff
@@ -168,11 +183,9 @@ class FilterFormat:
...
+    self.print_tests("INTERRUPTED TESTS", interruptions)
```
Comment (kwiberg) [inline gtest-parallel]: 1. Don't we already print failed tests?
2. When you print lists of tests, would it make sense to sort them first?

**Review by kwiberg — CHANGES_REQUESTED** [review]

**Comment by ehlemur-zz** [comment]
PTAL :smile:

**Comment by ehlemur-zz** [comment]
I think it's more readable the way it is.

**Review by kwiberg — APPROVED** [review]
lgtm

**Pull request merged**